deepcell_toolbox/utils.py

```diff
@@ -267,30 +267,69 @@ def resize(data, shape, data_format='channels_last'):
     is very fast. However, openCV does not support multi-channel resizing,
     so if the data has multiple channels, use skimage.
+
     Args:
-        data (np.array): data to be reshaped.
-        shape (tuple): shape of the output data.
+        data (np.array): data to be reshaped. Must have a channel dimension
+        shape (tuple): shape of the output data in the form (x,y).
+            Batch and channel dimensions are handled automatically and preserved.
         data_format (str): determines the order of the channel axis,
             one of 'channels_first' and 'channels_last'.
+
+    Raises:
+        ValueError: ndim of data not 3 or 4
+        ValueError: Shape for resize can only have length of 2, e.g. (x,y)
+
     Returns:
         numpy.array: data reshaped to new shape.
     """
+    if len(data.shape) not in {3, 4}:
+        raise ValueError('Data must have 3 or 4 dimensions, e.g. [batch, x, y], [x, y, channel]'
+                         'or [batch, x, y, channel]. Input data only has {} dimensions.'.format(
+                             str(len(data.shape))))
+
+    if len(shape) != 2:
+        raise ValueError('Shape for resize can only have length of 2, e.g. (x,y).'
+                         'Input shape has {} dimensions.'.format(str(len(shape))))
+
     # cv2 resize is faster but does not support multi-channel data
     # If the data is multi-channel, use skimage.transform.resize
     channel_axis = 0 if data_format == 'channels_first' else -1
+    batch_axis = -1 if data_format == 'channels_first' else 0
 
-    if data.shape[channel_axis] > 1:  # multichannel data, use skimage
-        # resize with skimage
+    # multichannel data, use skimage
+    if data.shape[channel_axis] > 1:
+        # Adjust output shape to account for channel axis
         if data_format == 'channels_first':
             shape = tuple([data.shape[channel_axis]] + list(shape))
         else:
             shape = tuple(list(shape) + [data.shape[channel_axis]])
-        resized = transform.resize(data, shape,
-                                   mode='constant',
-                                   preserve_range=True)
 
-    else:  # single channel image, resize with cv2
-        resized = cv2.resize(np.squeeze(data), shape)  # pylint: disable=E1101
+        # Check for batch dimension to loop over
+        if len(data.shape) == 4:
+            batch = []
+            for i in range(data.shape[batch_axis]):
+                d = data[i] if batch_axis == 0 else data[..., i]
+                rsz = transform.resize(d, shape, mode='constant', preserve_range=True)
+                batch.append(rsz)
+            resized = np.stack(batch, axis=batch_axis)
+        else:
+            resized = transform.resize(data, shape, mode='constant', preserve_range=True)
+
+    # single channel image, resize with cv2
+    else:
+        shape = tuple(shape)
+        # Check for batch dimension to loop over
+        if len(data.shape) == 4:
```

Comment:
This seems kinda fragile. What if it's length is 5?

Maybe there's a way to handle it generally, without essentially having 2 functions?

Reply:
I was struggling with this a bit and not loving all of the if/else statements. I picked length of 4 for the batch dimension because that is most common for our data ([batch, x, y, channel]).

I thought about actually splitting the code two functions where one handles the multichannel choice between skimage and cv2, while the other handles the batch dimension and calls the first function. Not sure if that is an actual improvement.

Reply:
I refactored the function to clean up the organization, but it still relies on the same underlying logic. Let me know what you think.

Reply:
There is a merge error in the code.

I am not sure, but I think there may be some unpredictability if the inptu data is rank 3.

Reply:
I did write tests for rank 3 data which seem to pass
```
# no batch, channel first
            in_shape = [c] + base_shape
            out_shape = tuple([c] + out)
            rs = utils.resize(np.random.rand(*in_shape), out, data_format='channels_first')
            assert out_shape == rs.shape

            # no batch, channel last
            in_shape = base_shape + [c]
            out_shape = tuple(out + [c])
            rs = utils.resize(np.random.rand(*in_shape), out, data_format='channels_last')
            assert out_shape == rs.shape
```

Reply:
What if there is batch, but no channel?

Reply:
I specified in the docstring that a channel dimension must be present. If you look back to the way the function was originally written it does assume that a channel dimension is present

Reply:
OK fair enough.